Fix e2e tests for price filter

imanish003 opened the pull request:
Some E2E tests related to the Price Filter block were failing. I have fixed the failing tests and made some other improvements. Here are the highlights of the changes:
1. **Refactored E2E Tests for Price Filter:** We have revamped our E2E tests in `price-filter.block_theme.side_effects.spec.ts`. A key highlight of this refactoring is the introduction of a new page object model, `PriceFilterPage`, which consolidates the logic for interacting with the price filter block. This includes adding the price filter block to new posts, setting the price filter range, and locating the images and products on a page.

2. **Modified Playwright Configuration:** Changes have been made to the Playwright configuration file `playwright.config.ts` to capture full-page screenshots only on failure. This helps in better understanding the failures and improves debugging efficiency.

**Impact of Changes**
- These changes will enhance the readability and maintainability of the code.
- It will also reduce the complexity of the test suite, making it easier to understand and modify.
- The improved Playwright configuration will assist in efficient debugging of test failures.

**Areas of Concern**
- Please review the changes in the `playwright.config.ts` to ensure that it aligns with our testing strategy.
- The refactored tests need to be carefully reviewed to ensure that all aspects of the previous tests are covered.
- The new `PriceFilterPage` class should be reviewed for potential improvements in the encapsulation of interactions with the price filter block.

Part of #9875 



### Testing

1. Run `npm run env:start` or `npm run env:restart`
2. Execute end-to-end tests with Playwright using the command: `npm run test:e2e-pw`.
3. Confirm that all tests complete successfully without any failures.
4. Verify that the Continuous Integration (CI) pipeline `Playwright Tests / Playwright E2E tests` runs without any errors or failures.

* [X] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [X] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

imanish003 committed:
Fix e2e tests for price filter

This commit makes two primary changes:

1. Update the Playwright configuration to capture full-page screenshots on test failures. This change will help in better debugging of failed tests by providing a full context of the application state at the time of the failure.

2. Refactor the end-to-end tests for the price filter feature of the WooCommerce Blocks plugin. The refactoring includes:
   - Introduction of a new `PriceFilterPage` class which encapsulates the logic for interacting with the price filter within the tests. This change improves the structure of the tests and makes them easier to understand.
   - The tests have been restructured to use this new class, resulting in a significant reduction in the amount of boilerplate code in the tests.
   - Removal of unused code and imports from the tests.

imanish003 committed:
Refactor price filter tests for clarity and remove unused utility file

Key changes:

1. Variable `img` was renamed to `firstProductImage` to improve clarity.

2. The method `setPrice` was renamed to `setPriceFilterRange` to better reflect its purpose.

3. Removed the function `urlSearchParamWhenFilterIsApplied` from the `blockData` object, as it was relocated to a private method inside the `PriceFilterPage` class named `generatePriceFilterQueryString`.

4. Replaced hardcoded strings 'woocommerce/filter-wrapper' and 'woocommerce/all-products' with constants WOOCOMMERCE_FILTER_WRAPPER and WOOCOMMERCE_ALL_PRODUCTS, respectively.

5. The `getAllProducts` method in `PriceFilterPage` class was renamed to `locateAllProducts` to reflect the actual operation it's performing.

6. The utility file `utils.ts` was deleted as it was no longer in use.

These changes make the test scripts more readable and maintainable, removing unnecessary code and improving method names for better understanding of their functionality.

kmanijak commented on tests/e2e-pw/tests/price-filter/price-filter.page.ts:
```diff
@@ -0,0 +1,173 @@
+		maxPrice: number | null
+	) {
+		let result = '';
+		if ( minPrice ) {
+			result += `min_price=${ minPrice }`;
+		}
+		if ( maxPrice ) {
+			result += `&max_price=${ maxPrice }`;
```
Detail, but in case there's no `min_price` query would start with an ampersand (`&max_price`). I don't know if that may influence the test results, but I'd keep it the closest to the real env possible. Unless you're sure that's ok.

imanish003 replied:
> Based on the output image it seems all good, but the event was not fired, so do you think it could be a race condition? 🤔

I am not sure because all tests are running successfully locally for me. I will try to reproduce the issue by following these steps:
1. Destroying the environment: `npm run wp-env destroy`
2. Start the environment again using: `npm run env:start`
3. Once the environment is up and running, re-run the tests using: `npm run test:e2e-pw`

I will update you once I have more information.

imanish003 replied:
> Detail, but in case there's no min_price query would start with an ampersand (&max_price). I don't know if that may influence the test results, but I'd keep it the closest to the real env possible. Unless you're sure that's ok.

Agreed, I made the changes in [8c3efcb](https://github.com/woocommerce/woocommerce-blocks/pull/10031/commits/8c3efcb206845d5ed68b8010e7d0b50e287dfc6d). Now function uses `URLSearchParams` to generate the query string. This object automatically encodes special characters and manages the & delimiter.

kmanijak reviewed (commented):
I reviewed the testing steps and they seem to cover the same as before. This PR really improved readability, so thanks for it!

For some reason, one test is failing for me locally:

```
1) [blockThemeWithGlobalSideEffects] › price-filter/price-filter.block_theme.side_effects.spec.ts:68:2 › woocommerce/price-filter Block - with PHP classic template › should show all products 

    Test timeout of 90000ms exceeded.
```
 It stops at:
```
      at price-filter/price-filter.page.ts:68
      > 68 |              await this.page.waitForLoadState( 'networkidle' );
```

Based on the output image it seems all good, but the event was not fired, so do you think it could be a race condition? 🤔 

![test-failed-1](https://github.com/woocommerce/woocommerce-blocks/assets/20098064/4321035e-8cd4-40b2-a5bf-bb12c190b261)


One minor thing, I know testing instructions won't be included in the release, but still worth adding a step to re-install deps to update `wp-env` locally before running the docker and tests.

kmanijak commented:
@imanish003, an update to the previous comment. I re-run the tests locally and this time two tests failed:

```
1) [blockThemeWithGlobalSideEffects] › price-filter/price-filter.block_theme.side_effects.spec.ts:68:2 › woocommerce/price-filter Block - with PHP classic template › should show all products 

 2) [blockThemeWithGlobalSideEffects] › price-filter/price-filter.block_theme.side_effects.spec.ts:75:2 › woocommerce/price-filter Block - with PHP classic template › should show only products that match the filter
```

on the exactly same step 

```
await this.page.waitForLoadState( 'networkidle' );
```

so seems like it introduces some flakiness 🤔 Do you think that may be related to my local setup somehow? Have you faced the same problem?

imanish003 committed:
Upgrade @playwright/test package

Upgrade `@playwright/test` package from version `1.32.3` to `1.35.1`.

kmanijak commented on package.json:
```diff
@@ -109,7 +109,7 @@
 		"@octokit/action": "5.0.2",
 		"@octokit/graphql": "5.0.5",
-		"@playwright/test": "1.32.3",
+		"@playwright/test": "^1.35.1",
```
We should stick to the fixed version:

```suggestion
		"@playwright/test": "1.35.1",
```

imanish003 replied:
I don't have any strong opinion on this as this is a dev dependency. I will go ahead and change it to fixed version 🤝

kmanijak replied:
I think that's the convention, but can't find the resource mentioning it, sorry. But I think we can assume that based on other dependencies 😅

imanish003 replied:
Made the changes in [ec0c613](https://github.com/woocommerce/woocommerce-blocks/pull/10031/commits/ec0c6132c23b85559d7d1c1b5a08a015c6e2071f)

imanish003 committed:
Change playwright version to fixed version

imanish003 committed:
Fix failing tests

imanish003 committed:
Wait for 1 second

imanish003 committed:
Refactor e2e tests for price filter and remove unnecessary timeouts

This commit performs a series of changes to the end-to-end tests related to the price filter:

1. Refactored the price filter e2e tests by splitting the `price-filter.block_theme.side_effects.spec.ts` test into two separate tests: `price-filter-with-all-products.block_theme.side_effects.spec.ts` and `price-filter-with-classic-template.block_theme.side_effects.spec.ts`. This provides better organization and clarity for these tests, and enables us to focus more closely on each distinct context in isolation.

2. Deleted the old `price-filter.block_theme.side_effects.spec.ts` test as its responsibilities have been taken over by the new tests.

3. Removed unnecessary `await page.waitForTimeout( 1000 );` lines from the classic-template block theme test and the products block theme test. This reduces the overall time taken to run these tests and eliminates a potential source of flakiness.

4. Made a small adjustment to the `price-filter.page.ts` by removing an unnecessary call to `this.editor.openDocumentSettingsSidebar()` and another unnecessary timeout.

imanish003 committed:
Refactor URL parameter creation in price filter test

This commit refactors the `getPriceFilterQuery` method in `price-filter.page.ts` to use the built-in URLSearchParams API for constructing URL parameters.

Previously, the method was manually creating a string representation of URL parameters. Now, it leverages the URLSearchParams API to improve readability and handle potential edge cases better.

The new approach appends 'min_price' and 'max_price' as URL parameters if they are not null. The `toString()` function is then called on the URLSearchParams object to return a string that can be used in a URL.

imanish003 committed:
Refactor E2E price filter tests and update test scripts in global-setup

Here are the key changes:
1. Refactored the way the 'wp action-scheduler run' command is run in the 'global-setup.ts' script. Now, it correctly separates the wp command from the arguments passed to it.
2. Consolidated the 'price-filter-with-all-products.block_theme.side_effects.spec.ts' and 'price-filter-with-classic-template.block_theme.side_effects.spec.ts' tests into a single 'price-filter.block_theme.side_effects.spec.ts' file. The new file contains tests for both scenarios: with all products block and with PHP classic template. Also, updated the product locator function to distinguish between classic templates and products beta block.
3. Deleted the original 'price-filter-with-all-products.block_theme.side_effects.spec.ts' and 'price-filter-with-classic-template.block_theme.side_effects.spec.ts' test files as they are no longer needed.
4. Updated the 'price-filter.page.ts' file to remove the 'waitForLoadState' method call after inserting the block in the editor and added logic to handle locating products in different scenarios (classic template vs products beta block).

imanish003 committed:
Use -- instead of quotes

imanish003 committed:
Minor improvements

imanish003 commented:
Hi @kmanijak, After making some changes. I have tested it locally & all tests are passing for me. Can you please review again? 

As you can see in the demo video below, all tests are passing for me:
`npm run test:e2e-pw -- --ui`

https://github.com/woocommerce/woocommerce-blocks/assets/16707866/5462cc95-ba8d-49c7-857e-aa31f9bdfc36

kmanijak reviewed (approved):
All the tests are passing just right! 💪
